docs/cswinrtgen/marshalling-arrays.md

```diff
@@ -0,0 +1,164 @@
+# CsWinRT 3.0 array infrastructure
+
+This document builds on top of the [generic interface marshalling infrastructure document](marshalling-generic-interfaces.md) and describes the design of the CsWinRT 3.0 array marshalling infrastructure. The same basic concepts outlined in the other document apply here, with the amin difference being a different API surface for marshaller types. This will be covered in more detail below.
+
+## Native -> managed (in `WinRT.Runtime.dll`)
+
+The runtime .dll will expose two infrastructure APIs for `WinRT.Interop.dll` to use for marshalling arrays (implementation specific).
+
+The first API is a new type of "`ComWrappers` callback", specialized for array types:
+
+```csharp
+public interface IWindowsRuntimeArrayComWrappersCallback
+{
+    static abstract Array CreateArray(uint count, void* value);
+}
+```
+
+And a marshaller type to support this:
+
+```csharp
+public static class WindowsRuntimeArrayMarshaller
+{
+    public static Array? ConvertToManaged<TCallback>(void* value, in Guid iid)
+        where TCallback : IWindowsRuntimeArrayComWrappersCallback, allows ref struct;
+}
+```
+
+The `ConvertToManaged` method will perform a `QueryInterface` on the input `value` pointer for the specified IID (as this method is only used to unbox arrays that are passed as opaque objects). It will then call `Value` on the ```IReferenceArray`1<T>``` interface pointer, and invoke the supplied callback to return the unboxed managed array instance.
+
+## Native -> managed (in `WinRT.Interop.dll`)
+
+Each array type will have a corresponding marshaller callback type, implementing `IWindowsRuntimeArrayComWrappersCallback`. Note that inputs to the `CreateArray` method will always be not-null. The generated `<string>ArrayMarshaller` type is detailed at the end of this document. This is what a generated callback type would look like for a `string` array:
+
+```csharp
+public abstract class <string>ArrayComWrappersCallback : IWindowsRuntimeArrayComWrappersCallback
```

Comment:
Generate as `static` instead of `abstract` if it isn't supposed to be implemented? 
```suggestion
public static class <string>ArrayComWrappersCallback : IWindowsRuntimeArrayComWrappersCallback
```

Reply:
I would love to do that, but I can't make it static since I'm passing it as a type argument 🥲

Making it statis was a suggestion from @MichalStrehovsky to minimize the metadata being kept, as at least this way ILC can guarantee the type will never be constructed. I think it's the closest I can get for now until .NET lifts this very annoying restriction. _In theory_ I could emit it as static given I'm emitting IL, but I would prefer to not emit IL that Roslyn would never generate, to avoid issues with CoreCLR/NAOT seeing stuff it doesn't expect. Does that make sense?

Reply:
Right, C# never added the "static classes in generics" feature. Nevermind then.